stdenv.mkDerivation: check that depsBuildX are executable on the build system

Artturin opened the pull request:
```
nix-repl> :p pkgsCross.aarch64-multiplatform.keybinder.nativeBuildInputs
[ «derivation /nix/store/sxxk0g9gd18w9r28gbfx8wl65sr8liwm-aarch64-unknown-linux-gnu-pkg-config-wrapper-0.29.2.drv» «derivation /nix/store/ldbs4pw57hy3szqfh8ycskrgywpaahqs-autoconf-2.71.drv» trace: warning: automake-aarch64-unknown-linux-gnu-1.11.6 in keybinder nativeBuildInputs not executable on build
«derivation /nix/store/yj8vj3gbd78scrj4xjmdx4rm5p87dyb9-automake-aarch64-unknown-linux-gnu-1.11.6.drv» ]
```

TODO less duplication and better warning by merging with checkDependencyList/make a function which takes these as args

SuperSandro2000 commented:
There seems to be a lot of errors around i686-w64-mingw32-stage-final-gcc-wrapper but I like the general idea.

Artturin commented:
```
nix-repl> o = lib.lists.last wine.nativeBuildInputs

nix-repl> o
«derivation /nix/store/i7ya4lqal0awbgxrabkq70gx56skvax5-i686-w64-mingw32-stage-final-gcc-wrapper-11.3.0.drv»

nix-repl> stdenv.buildPlatform.canExecute o.stdenv.hostPlatform
true

nix-repl> pkgsi686Linux.stdenv.buildPlatform.canExecute o.stdenv.hostPlatform
false
```

Artturin commented:
this fixes the eval error, i copied it from wineWow
```diff
diff --git a/pkgs/applications/emulators/wine/packages.nix b/pkgs/applications/emulators/wine/packages.nix
index 3b5aa19658c..440fd167d98 100644
--- a/pkgs/applications/emulators/wine/packages.nix
+++ b/pkgs/applications/emulators/wine/packages.nix
@@ -11,6 +11,7 @@ in with src; {
   wine32 = pkgsi686Linux.callPackage ./base.nix {
     pname = "wine";
     inherit src version supportFlags patches moltenvk;
+    stdenv = stdenv_32bit;
     pkgArches = [ pkgsi686Linux ];
     vkd3dArches = lib.optionals supportFlags.vkd3dSupport [ vkd3d_i686 ];
     geckos = [ gecko32 ];
```

but it does change
```
checking build system type... i686-pc-linux-gnu
checking host system type... i686-pc-linux-gnu
```

to

```
wine> checking build system type... x86_64-pc-linux-gnu
wine> checking host system type... x86_64-pc-linux-gnu
```

```
# A stdenv capable of building 32-bit binaries.
# On x86_64-linux, it uses GCC compiled with multilib support; on i686-linux,
# it's just the plain stdenv.
stdenv_32bit = lowPrio (if stdenv.hostPlatform.is32bit then stdenv else multiStdenv);
```

i don't know if this affects anything negatively but wine builds and the log output is similar but with stuff like `gcc -m32 ...` and `-m32` in other places

Weird things in the log are

```diff
- checking for -lcups... libcups.so.2
+ checking for -lcups... not found

- configure: libcapi20 development files not found, ISDN won't be supported.
+ configure: libcapi20 32-bit development files not found, ISDN won't be supported.
+ configure: libcups 32-bit development files not found, CUPS won't be supported.

Artturin committed:
wine: use stdenv_32bit for wine32

to make sure all things in nativeBuildInputs are executable on
buildPlatform

```
checking build system type... i686-pc-linux-gnu
checking host system type... i686-pc-linux-gnu
```

to

```
wine> checking build system type... x86_64-pc-linux-gnu
wine> checking host system type... x86_64-pc-linux-gnu
```

i don't know if this affects anything negatively but wine builds and the log output is similar but with stuff like `gcc -m32 ...` and `-m32` in other places

Weird things in the log are

```diff
checking for cups/cups.h... yes
checking for cups/ppd.h... yes
- checking for -lcups... libcups.so.2
+ checking for -lcups... not found
...
- configure: libcapi20 development files not found, ISDN won't be supported.
+ configure: libcapi20 32-bit development files not found, ISDN won't be supported.
+ configure: libcups 32-bit development files not found, CUPS won't be supported.
```

no diff in
```
nix-repl> pkgs = import ./. { system = "i686-linux"; }

nix-repl> pkgs.wine
«derivation /nix/store/lik48akxk5hk3hvbqhaxgsh0827kv3zd-wine-7.0.drv»
```

Artturin committed:
mfcl3770cdw: use stdenv_32bit instead of pkgsi686Linux

 ["trace: warning: dpkg-1.21.1ubuntu2.1 in mfcl3770cdwdrv-1.0.2-0 nativeBuildInputs not executable on build", "trace: warning: hook in mfcl3770cdwdrv-1.0.2-0 nativeBuildInputs not executable on build", "trace: warning: dpkg-1.21.1ubuntu2.1 in mfcl3770cdwdrv-1.0.2-0 nativeBuildInputs not executable on build", "trace: warning: hook in mfcl3770cdwdrv-1.0.2-0 nativeBuildInputs not executable on build"]

```
nix-repl> o = lib.lists.last wine.nativeBuildInputs

nix-repl> o
«derivation /nix/store/i7ya4lqal0awbgxrabkq70gx56skvax5-i686-w64-mingw32-stage-final-gcc-wrapper-11.3.0.drv»

nix-repl> stdenv.buildPlatform.canExecute o.stdenv.hostPlatform
true

nix-repl> pkgsi686Linux.stdenv.buildPlatform.canExecute o.stdenv.hostPlatform
false
```

Artturin committed:
pkgs/stdenv/generic/make-derivation.nix: run nixpkgs-fmt

allows us to use nixpkgs-fmt when editing this file making the editing
experience much better

Artturin committed:
stdenv.mkDerivation: check that depsBuildX are executable on the build system

```
nix-repl> pkgsCross.aarch64-multiplatform.keybinder.nativeBuildInputs
[ «derivation /nix/store/sxxk0g9gd18w9r28gbfx8wl65sr8liwm-aarch64-unknown-linux-gnu-pkg-config-wrapper-0.29.2.drv» «derivation /nix/store/ldbs4pw57hy3szqfh8ycskrgywpaahqs-autoconf-2.71.drv» trace: warning: automake-aarch64-unknown-linux-gnu-1.11.6 in keybinder's nativeBuildInputs not executable on build
«derivation /nix/store/yj8vj3gbd78scrj4xjmdx4rm5p87dyb9-automake-aarch64-unknown-linux-gnu-1.11.6.drv» ]
```

Artturin commented:
lots of warnings in perlPackages and pythonPackages
pythonPackages fix in #201734 

perlPackages todo for me, will be fixed before this pr is merged or in this pr

Artturin commented on pkgs/development/interpreters/perl/default.nix:
```diff
@@ -233,24 +234,21 @@ let
   perl536 = common {
-    perl = pkgs.perl536;
-    buildPerl = buildPackages.perl536;
+    self = __splicedPackages.perl536;
```
preparation for doing splicing like in python

Artturin committed:
stdenv.mkDerivation: workaround hacks in perl cross-compilation

Artturin committed:
perl: use self == __splicedPackages instead of perl and buildPerl

Artturin commented:
somehow the mini output of perl gets introduced 

before

```
nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.perlPackages.perl.mini
error: attribute 'mini' missing

       at «string»:1:1: (source not available)
       Did you mean man?

nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.perlPackages.perl
«derivation /nix/store/bjh5sr9f1g48q1by44qnabgkdiayi5h9-perl-5.36.0.drv»
```

after
```
nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.perlPackages.perl.mini
«derivation /nix/store/x873z4mmfqwynw10a8j5v9xzm8mharsk-perl-aarch64-unknown-linux-gnu-5.36.0.drv»

nix-repl> pkgsCross.aarch64-multiplatform.buildPackages.perlPackages.perl
«derivation /nix/store/bjh5sr9f1g48q1by44qnabgkdiayi5h9-perl-5.36.0.drv»
```

Artturin committed:
stdenv.mkDerivation: workaround

amjoseph-nixpkgs commented:
You can ping me for a review when this is ready.
